clients/drcachesim/tests/trace_interval_analysis_unit_tests.cpp

```diff
@@ -149,63 +340,193 @@ class test_analysis_tool_t : public analysis_tool_t {
+        // does not guarantee that workers will process shards one after the other.)
+        // Expected active interval_id: shard_1_local - shard_2_local - whole_trace
+        gen_marker(1, TRACE_MARKER_TYPE_TIMESTAMP, 40),  // 0 - _ - 0
+        gen_instr(1, 10000),                             // 0 - _ - 0
+        gen_data(1, true, 1234, 4),                      // 0 - _ - 0
+        gen_marker(2, TRACE_MARKER_TYPE_TIMESTAMP, 151), // _ - 0 - 1
+        gen_instr(2, 20000),                             // _ - 0 - 1
+        gen_marker(1, TRACE_MARKER_TYPE_TIMESTAMP, 170), // 1 - _ - 1
+        gen_instr(1, 10008),                             // 1 - _ - 1
+        gen_marker(1, TRACE_MARKER_TYPE_TIMESTAMP, 201), // 2 - _ - 2
+        gen_instr(1, 20004),                             // 2 - _ - 2
+        gen_marker(2, TRACE_MARKER_TYPE_TIMESTAMP, 210), // _ - 1 - 2
+        gen_instr(2, 20008),                             // _ - 1 - 2
+        gen_marker(2, TRACE_MARKER_TYPE_TIMESTAMP, 270), // _ - 1 - 2
+        gen_instr(2, 20008),                             // _ - 1 - 2
+        gen_marker(2, TRACE_MARKER_TYPE_TIMESTAMP, 490), // _ - 3 - 4
+        gen_instr(2, 20012),                             // _ - 3 - 4
+        gen_marker(1, TRACE_MARKER_TYPE_TIMESTAMP, 590), // 5 - _ - 5
+        gen_exit(1),                                     // 5 - _ - 5
+        gen_marker(2, TRACE_MARKER_TYPE_TIMESTAMP, 610), // _ - 5 - 6
+        gen_instr(2, 20016),                             // _ - 5 - 6
+        gen_exit(2)                                      // _ - 5 - 6
+    };
+    // Format for interval_switch_point: <tid, seen_memrefs, interval_id>
+    // For serial: each whole trace interval is made up of only one snapshot, the
+    // serial snapshot.
+    std::vector<test_analysis_tool_t::recorded_snapshot_t> serial_state_snapshots = {
+        test_analysis_tool_t::recorded_snapshot_t(0, 0, 100, { { 0, 3, 0 } }),
+        test_analysis_tool_t::recorded_snapshot_t(0, 1, 200, { { 0, 7, 1 } }),
+        test_analysis_tool_t::recorded_snapshot_t(0, 2, 300, { { 0, 13, 2 } }),
+        test_analysis_tool_t::recorded_snapshot_t(0, 4, 500, { { 0, 15, 4 } }),
+        test_analysis_tool_t::recorded_snapshot_t(0, 5, 600, { { 0, 17, 5 } }),
+        test_analysis_tool_t::recorded_snapshot_t(0, 6, 700, { { 0, 20, 6 } }),
+    };
+    // For parallel: each whole trace interval is made up of snapshots from each
+    // shard that was active in that interval.
+    std::vector<test_analysis_tool_t::recorded_snapshot_t> parallel_state_snapshots = {
+        test_analysis_tool_t::recorded_snapshot_t(0, 0, 100, { { 1, 3, 0 } }),
+        // Narration: The whole-trace interval_id=1 with interval_end_timestamp=200
+        // is made up of the following two shard-local interval snapshots:
+        // - from shard_id=1, the interval_id=1 that ends at the local_memref=5
+        // - from shard_id=2, the interval_id=0 that ends at the local_memref=2
```

Comment:
Wait, how can interval 0 be combined with interval 1 if you're doing (1) when merging: ignoring inactive shards?

Reply:
The interval_id in component_intervals is shard-local. The final whole-trace interval id is the other argument to recorded_snapshot_t.

Reply:
Since tid=52 starts later, its shard-local interval_ids don't align with the other thread's.  But note that the interval_end_timestamp is what we use to decide what to merge together.

Reply:
I thought you changed the interval id to be computed from the absolute timestamp, so that they would all line up.

Reply:
See the comment at clients/drcachesim/analyzer.cpp:308 which talks about starting interval ids at zero for each shard.

Reply:
We want the interval boundary (the end timestamps) to line up, not the ids themselves.